oak_runtime/src/node/http/mod.rs

```diff
@@ -481,12 +482,31 @@ impl Pipe {
 }
 
 fn get_oak_label(req: &HttpRequest) -> Result<Label, OakStatus> {
-    match req.headers.get(oak_abi::OAK_LABEL_HTTP_KEY) {
-        Some(label_str) => parse_json_label(label_str.to_vec()),
-        None => {
-            warn!("No HTTP label found.");
-            Err(OakStatus::ErrInvalidArgs)
-        }
+    let json_label = req.headers.get(oak_abi::OAK_LABEL_HTTP_JSON_KEY);
+    let protobuf_label = req.headers.get(oak_abi::OAK_LABEL_HTTP_PROTOBUF_KEY);
+
+    if json_label.is_some() && protobuf_label.is_some()
+        || json_label.is_none() && protobuf_label.is_none()
+    {
+        error!(
+            "Exactly one header must be provided as an {} or {} header.",
+            oak_abi::OAK_LABEL_HTTP_JSON_KEY,
+            oak_abi::OAK_LABEL_HTTP_PROTOBUF_KEY
+        );
+        return Err(OakStatus::ErrInvalidArgs);
+    }
+
+    if let Some(label_str) = json_label {
+        parse_json_label(label_str.to_vec())
+    } else if let Some(label_bytes) = protobuf_label {
+        Label::decode(&label_bytes[..]).map_err(|err| {
+            warn!("Could not parse HTTP label: {}", err);
+            OakStatus::ErrInvalidArgs
+        })
+    } else {
+        // Should never arrive here
+        warn!("No HTTP label found.");
+        Err(OakStatus::ErrInvalidArgs)
     }
```

Comment:
I'm still new to Rust, but it seems to me this could be restructured into a match expression? 

```rust
fn get_oak_label(req: &HttpRequest) -> Result<Label, OakStatus> {
    let headers = (
        req.headers.get(oak_abi::OAK_LABEL_HTTP_JSON_KEY),
        req.headers.get(oak_abi::OAK_LABEL_HTTP_PROTOBUF_KEY),
    );

    match headers {
        (Some(json_label), None) => parse_json_label(json_label.to_vec()),
        (None, Some(protobuf_label)) => Label::decode(&protobuf_label[..]).map_err(|err| {
            warn!("Could not parse HTTP label: {}", err);
            OakStatus::ErrInvalidArgs
        }),
        _ => {
            error!(
                "Exactly one header must be provided as an {} or {} header.",
                oak_abi::OAK_LABEL_HTTP_JSON_KEY,
                oak_abi::OAK_LABEL_HTTP_PROTOBUF_KEY
            );
            Err(OakStatus::ErrInvalidArgs)
        }
    }
}
```

Reply:
I really like this. Thanks :)

Reply:
Rust's pattern matching is quite powerful, I really miss it when writing code in other languages now.